src/sentry/tasks/app_store_connect.py

```diff
@@ -39,44 +40,70 @@ def inner_dsym_download(
-        except AppConnectBuild.DoesNotExist:
-            build_state = AppConnectBuild(
-                project=project,
-                app_id=build.app_id,
-                bundle_id=config.bundleId,
-                platform=build.platform,
-                bundle_short_version=build.version,
-                bundle_version=build.build_number,
-                fetched=False,
-            )
 
+    # persist all fetched builds into the database as "pending"
+    builds = []
+    for build in client.list_builds():
+        build_state = get_or_create_persisted_build(project, config, build)
         if not build_state.fetched:
-            with tempfile.NamedTemporaryFile() as dsyms_zip:
-                try:
-                    itunes_client.download_dsyms(build, pathlib.Path(dsyms_zip.name))
-                except appconnect.NoDsymsError:
-                    logger.debug("No dSYMs for build %s", build)
-                    continue
+            builds.append(build_state)
+
+    itunes_client = client.itunes_client()
+    for build_state in builds:
+        with tempfile.NamedTemporaryFile() as dsyms_zip:
+            try:
+                itunes_client.download_dsyms(build, pathlib.Path(dsyms_zip.name))
+
                 create_difs_from_dsyms_zip(dsyms_zip.name, project)
-            build_state.fetched = True
-            build_state.save()
-            logger.debug("Uploaded dSYMs for build %s", build)
+                logger.debug("Uploaded dSYMs for build %s", build)
+            except appconnect.NoDsymsError:
+                logger.debug("No dSYMs for build %s", build)
+            except ITunesSessionExpiredException:
+                logger.debug("Error fetching dSYMs: expired iTunes session")
+                # we early-return here to avoid trying all the other builds
+                # as well, since an expired token will error for all of them.
+                # we also swallow the error and not report it because this is
+                # a totally expected error and not actionable.
+                return
+
+        # If we either downloaded, or didn't need to download the dSYMs
+        # (there was no dSYM url), we check off this build.
+        build_state.fetched = True
+        build_state.save()
 
 
 def create_difs_from_dsyms_zip(dsyms_zip: str, project: Project) -> None:
     with open(dsyms_zip, "rb") as fp:
         created = debugfile.create_files_from_dif_zip(fp, project, accept_unknown=True)
         for proj_debug_file in created:
             logger.debug("Created %r for project %s", proj_debug_file, project.id)
+
+
+def get_or_create_persisted_build(
+    project: Project, config: appconnect.AppStoreConnectConfig, build: appconnect.BuildInfo
+) -> AppConnectBuild:
+    """Fetches the sentry-internal :class:`AppConnectBuild` corresponding
+    to the provided :class:`appconnect.BuildInfo` as returned by the
+    AppStore Connect API. If no build exists yet, a new "pending" build is
+    created."""
```

Comment:
definitely a nit, but I really like one-line summaries as docstrings, I believe this is even in both the python and rust style guides.